src/ac_power_flow_jacobian.jl

```diff
@@ -1,228 +1,511 @@
-                    J[F_ix_f_i, X_ix_t_fst] =
-                        -2 * imag(Yb[ix_f, ix_t]) * Vm[ix_f] + sum(
-                            Vm[k] * (
-                                real(Yb[ix_f, k]) * sin(θ[ix_f] - θ[k]) -
-                                imag(Yb[ix_f, k]) * cos(θ[ix_f] - θ[k])
-                            ) for k in data.neighbors[ix_f] if k != ix_f
-                        )
-                    #Jac: Reactive PF against same angle θ[ix_f]
-                    J[F_ix_f_i, X_ix_t_snd] =
-                        Vm[ix_f] * sum(
-                            Vm[k] * (
-                                real(Yb[ix_f, k]) * cos(θ[ix_f] - θ[k]) -
-                                imag(Yb[ix_f, k]) * -sin(θ[ix_f] - θ[k])
-                            ) for k in data.neighbors[ix_f] if k != ix_f
-                        )
-                else
-                    g_ij = real(Yb[ix_f, ix_t])
-                    b_ij = imag(Yb[ix_f, ix_t])
-                    #Jac: Active PF w/r to different voltage magnitude Vm[ix_t]
-                    J[F_ix_f_r, X_ix_t_fst] =
-                        Vm[ix_f] *
-                        (g_ij * cos(θ[ix_f] - θ[ix_t]) + b_ij * sin(θ[ix_f] - θ[ix_t]))
-                    #Jac: Active PF w/r to different angle θ[ix_t]
-                    J[F_ix_f_r, X_ix_t_snd] =
-                        Vm[ix_f] *
-                        Vm[ix_t] *
-                        (g_ij * sin(θ[ix_f] - θ[ix_t]) + b_ij * -cos(θ[ix_f] - θ[ix_t]))
-
-                    #Jac: Reactive PF w/r to different voltage magnitude Vm[ix_t]
-                    J[F_ix_f_i, X_ix_t_fst] =
-                        Vm[ix_f] *
-                        (g_ij * sin(θ[ix_f] - θ[ix_t]) - b_ij * cos(θ[ix_f] - θ[ix_t]))
-                    #Jac: Reactive PF w/r to different angle θ[ix_t]
-                    J[F_ix_f_i, X_ix_t_snd] =
-                        Vm[ix_f] *
-                        Vm[ix_t] *
-                        (g_ij * -cos(θ[ix_f] - θ[ix_t]) - b_ij * sin(θ[ix_f] - θ[ix_t]))
-                end
-            else
-                error("Undefined Conditional")
-            end
+    for bus_from in 1:num_buses
+        row_from_p = 2 * bus_from - 1
+        row_from_q = 2 * bus_from
+
+        for bus_to in data.neighbors[bus_from]
+            col_to_vm = 2 * bus_to - 1
+            col_to_va = 2 * bus_to
+            bus_type = data.bus_type[bus_to, time_step]
+            _set_entries_for_neighbor(Jv, Yb, Vm, θ,
+                bus_from, bus_to,
+                row_from_p, row_from_q,
+                col_to_vm, col_to_va,
+                data.neighbors[bus_from], Val(bus_type))
         end
     end
     return
 end
+
+function calculate_loss_factors(
+    data::ACPowerFlowData,
+    Jv::SparseMatrixCSC{Float64, Int32},
+    time_step::Int,
+)
+    num_buses = first(size(data.bus_type))
+    ref, pv, pq = bus_type_idx(data, time_step)
+    pvpq = vcat(pv, pq)
+    pvpq_coords = [
+        x for pair in zip(
```

Comment:
Is there a way to rewrite this more legibly using functions like `collect`, `vcat`, `map` instead of the two-level list comprehension?

Reply:
I integrated the function penalty_factors! into calculate_loss_factors and simplified the code quite a bit:

https://github.com/NREL-Sienna/PowerFlows.jl/blob/c12da85c26fa3097d94b9166dc1fe25ac2d532b7/src/ac_power_flow_jacobian.jl#L508

Reply:
also seems like the windows error got resolved